### ISSUE TITLE
"reference has changed concurrently" when fetch and checkout.

### ISSUE THREAD

**Issue by miltalex** [go-git/go-git#37]
When I try to do a git fetch and then a git checkout to a specific revision I get the following error. 

`error: reference has changed concurrently`

It seems that after 2 retries the whole fetch checkout operation works fine.

There was an old similar issue [here](https://github.com/src-d/go-git/issues/1230).

**Comment by mcuadros** [comment]
Are you using go routines?

**Comment by miltalex** [comment]
No. I just do a git fetch in the beggining.
```
err = repo.Fetch(&git.FetchOptions{
	RemoteName: "origin",
	Auth:       r.Auth,
})
```
Then I resolve the reference that user provided, by checking `referenceTag` and `referenceBranch`, and then I do a checkout and a pull.
```
err = w.Checkout(&git.CheckoutOptions{
	Hash:  *revision,
	Force: true,
})

err = w.Pull(&git.PullOptions{
	RemoteName:    "origin",
	ReferenceName: plumbing.ReferenceName(r.ReferenceName),
	Auth:          r.Auth,
}) 
```

A summary of the code.

**Comment by mcuadros** [comment]
Can you create a script that reproduces the bug, to be able to take a look?

**Comment by miltalex** [comment]
Unfortunately not I can share the code. I noticed that this will happen if something changed on the remote when we call the function, but I wasn't able to reproduce it.

**Comment by cumminsd** [comment]
I also am getting the issue occasionally. It seems to be breaking references:

`error: cannot lock ref 'refs/remotes/origin/EVO-26728': unable to resolve reference 'refs/remotes/origin/EVO-26728': reference broken`

It does appear to be when someone else changes the repository. Here is the function I am calling to update the repository. Note that the commented out section was used to check the repository before calling update, but I entered an issue for that. (If there is an empty folder in the repository, worktree.Status() returns that the tree is dirty, if when it isn't.)

`
// Update does a pull on the specified repository
func Update(folder string) ([]*object.Commit, error) {
	repo, err := git.PlainOpen(folder)
	if err != nil {
		return nil, fmt.Errorf("Error opening repository: %s", err)
	}

	auth := githttp.BasicAuth{Username: "xxxxx", Password: "xxxxx"}

	oldHead, err := repo.Head()
	if err != nil {
		return nil, fmt.Errorf("Error finding HEAD: %s", err)
	}

	oldCommit, err := repo.CommitObject(oldHead.Hash())
	if err != nil {
		return nil, fmt.Errorf("Error retrieving last commit: %s", err)
	}

	workTree, err := repo.Worktree()
	if err != nil {
		return nil, fmt.Errorf("Error opening work tree: %s", err)
	}

	// status, err := workTree.Status()
	// if err != nil {
	// 	if err == git.ErrNonFastForwardUpdate {
	// 		return nil, fmt.Errorf("Project needs to be pushed")
	// 	}
	// 	return nil, fmt.Errorf("Error getting status: %s", err)
	// }

	// if !status.IsClean() {
	// 	return nil, fmt.Errorf("Project has uncommitted changes")
	// }

	err = workTree.Pull(&git.PullOptions{
		Auth: &auth,
	})
	if err != nil {
		if err == git.NoErrAlreadyUpToDate {
			return nil, nil
		}
		return nil, fmt.Errorf("Error pulling project: %s", err)
	}

	logs := []*object.Commit{}

	lgs, err := repo.Log(&git.LogOptions{})
	if err != nil {
		return nil, fmt.Errorf("Error retrieving logs: %s", err)
	}

	lgs.ForEach(func(c *object.Commit) error {
		if c.Author.When.After(oldCommit.Author.When) {
			logs = append(logs, c)
		}
		return nil
	})

	return logs, nil
}
`

**Comment by sitilge** [comment]
@mcuadros I am also experiencing the same problem with a large repo:

```go

var gitRepo *git.Repository
var gitWorktree *git.Worktree

func init() {
        var err error
	gitRepo, err = git.PlainOpen(localRoot)
	//handleError("git.PlainOpen", err)

	gitWorktree, err = gitRepo.Worktree()
	//handleError("git.Worktree", err)
}
```
and then I call `Pull` in another func (no goroutines)

```go
err := gitWorktree.Pull(&git.PullOptions{
	RemoteName: "origin",
})
if err != nil {
	return err
}
```
It is strange but when I do manually `git pull`, the error disappears when I run the go program.

**Comment by ldrmcml** [comment]
In [kubernetes/publishing-bot](https://github.com/kubernetes/publishing-bot) also encounter this problem:  
https://github.com/kubernetes/publishing-bot/blob/ba0e69a186c2848c1719509b702a821c8a83ef80/cmd/publishing-bot/main.go#L188

**Comment by lostiniceland** [comment]
Seing the same problem when doing a PlainOpen on an existing local repo and issuing a Fetch

**Comment by basyas** [comment]
i got same problem too

**Comment by icecoffee531** [comment]
+1

**Comment by brighten1023** [comment]
I found the solution. An empty file that cannot be removed in .git/refs/remotes/origin since not enough time between checkout and pull operation. The solution is sleep goroutine after the checkout operation.
```
err = w.Checkout(&git.CheckoutOptions{
	Hash:  *revision,
	Force: true,
})

time.Sleep(2*time.Second)

err = w.Pull(&git.PullOptions{
	RemoteName:    "origin",
	ReferenceName: plumbing.ReferenceName(r.ReferenceName),
	Auth:          r.Auth,
})
 ```

**Comment by brandonsturgeon** [comment]
This happens to me too, I'm just doing something like:
```go
repo, err := git.PlainOpen(repoPath)
w, err := repo.Worktree()
status, err := w.Status()

if status.IsClean() {
    err = w.Pull(&git.PullOptions{
        RemoteName: "origin"
    })
}
```

I get the status of the given repo, and if it's clean, I attempt to pull. So not checking out but just getting the local status.

**Comment by arloliu** [comment]
This happens to me too. 

The reproduction way is removing a folder containing many files in repo after repo opened and before checkout. Then this issue 100% appears

I found a workaround.
After `PlainOpen` or work tree `Checkout`. Sometimes the work tree status is not clean.
But if I execute work tree Reset again (HardReset), then it can pull successfully.

Example:
```go
repo, err := git.PlainOpen(repoPath)
w, err := repo.Worktree()

// remove a a folder containing many files

w.Checkout(&git.CheckoutOptions{Branch: refName, Create: false, Force: true, Keep: false})

status, err := w.Status()

if !status.IsClean() {
    ref, _ := repo.Head()
    w.Reset(&git.ResetOptions{Commit: ref.Hash(), Mode: git.HardReset})
}

w.Pull(&git.PullOptions{...})

```

**Comment by akhilerm** [comment]
We are hitting this bug frequently in kubernetes.
Ref : https://github.com/kubernetes/kubernetes/issues/56876#issuecomment-1704960096

Retrying after sometime solves it. But sometimes it take more no.of retries, so that we do not hit the issue. 

We are not using any goroutines to concurrently do any operation on the repository.

Does repository size have a relation with the issue?

**Comment by pjbgf** [comment]
We are still trying pin-point the root cause of this. For that, it would be nice to confirm whether this is only the filesystem dotgit storage or if this also affects the in memory implementation. If you are using the latter (as per example below), please add a :+1: to this comment or reply with more details on your implementation/outcome:

```golang
git.Clone(memory.NewStorage(), nil, &git.CloneOptions{
		URL: "https://github.com/<REPO>",
	})
```

Please note that anyone using `git.PlainClone` are automatically using the filesystem-based dotgit storage.

**Comment by Narrat** [comment]
As someone from issue #314 who is only affected from using `Fetch` on regular git-scm cloned repositories.
The in-memory implementation. Is it comparable to a `tmpfs`? If so, such test with such an environment fairly easy (but I suspect it is to different).
If not I may the able to adjust my small program to include cloning a repo and waiting some time before fetching.

**Comment by Javier-varez** [comment]
I ran into this problem recently. I found a repository that consistently could not fetch references. Initially it failed with this message:
```
error: reference has changed concurrently
```

And in subsequent fetch attempts it failed with:
```
error: some refs were not updated
```

I investigated the issue, and it looks like the problem is related to packed references not being correctly handled. The issue can be reproduced  when:
- A repository has a packed reference that was updated in the remote, but the local `.git/refs/...` file does not exist (i.e.: the reference exists only in the `.git/packed-refs` file.
- A fetch operation is attempted with the current `go-git` master branch.

The following script creates a minimal repo setup that reproduces the problem.
```bash
#!/bin/bash -e

PARENT_FOLDER="git"

# Local 1 will become the broken repo
REMOTE_DIR="$PWD/$PARENT_FOLDER/remote"
LOCAL1_DIR="$PWD/$PARENT_FOLDER/broken_repo"
LOCAL2_DIR="$PWD/$PARENT_FOLDER/another_local"

git init --bare ${REMOTE_DIR}

git clone ${REMOTE_DIR} ${LOCAL1_DIR}
git clone ${REMOTE_DIR} ${LOCAL2_DIR}

# Set up the inital repor in local1
git -C ${LOCAL1_DIR} commit -m "test1" --allow-empty
git -C ${LOCAL1_DIR} commit -m "test2" --allow-empty
git -C ${LOCAL1_DIR} update-ref refs/namespace/test_ref HEAD

# Publish updates to the remote
git -C ${LOCAL1_DIR} push origin "refs/namespace/*:refs/namespace/*"

# Now fetch and pach the local refs
git -C ${LOCAL1_DIR} fetch origin "refs/namespace/*:refs/remotes/origin/namespace/*"
git -C ${LOCAL1_DIR} pack-refs --all
# At this point local 1 has a refs/remotes/origin/namespace/test_ref reference that only exists
# in the pack-refs file

# On local2, create updates that local1 does not have and update the reference
git -C ${LOCAL2_DIR} fetch origin "refs/namespace/*:refs/namespace/*"
git -C ${LOCAL2_DIR} checkout refs/namespace/test_ref
git -C ${LOCAL2_DIR} commit -m "test3" --allow-empty
git -C ${LOCAL2_DIR} update-ref refs/namespace/test_ref HEAD

# Publish local2 updates to the remote
git -C ${LOCAL2_DIR} push origin "refs/namespace/*:refs/namespace/*"
```

And running this snippet with `go-git` triggers the bug:
```go
refspec := "refs/namespace/*:refs/remotes/origin/namespace/*"
r, err := git.PlainOpen("<path-to-broken-repo>")
CheckIfError(err)

err = r.Fetch(&git.FetchOptions{RemoteName: "origin", RefSpecs: []config.RefSpec{config.RefSpec(refspec)}})
CheckIfError(err)
```

I tracked the problem down to an improper handling of the `SetRef` function in https://github.com/go-git/go-git/blob/55f5d832fad510fecbad3fb413875d969e1441fb/storage/filesystem/dotgit/dotgit_setref.go#L45, which does not handle the case where the reference exists, but the file does not (as it is only present in the `packed-refs`.

I will submit a PR with a fix for this issue shortly.

**Comment by pjbgf** [comment]
Closing this as per merged #1116.

**Comment by joonas-fi** [comment]
@Javier-varez thank you so much for your expert detective work in fixing this! :heart: We at my work have been bitten by this bug so many times, I'm very glad to see this fixed!

Awesome, awesome work.
